google/cloud/bigtable/mutations_batcher.py

```diff
@@ -46,61 +177,318 @@ class MutationsBatcher:
+        batch_operation_timeout: float | None = None,
+        batch_per_request_timeout: float | None = None,
     ):
-        raise NotImplementedError
+        """
+        Args:
+          - table: Table to preform rpc calls
+          - flush_interval: Automatically flush every flush_interval seconds.
+              If None, no time-based flushing is performed.
+          - flush_limit_mutation_count: Flush immediately after flush_limit_mutation_count
+              mutations are added across all entries. If None, this limit is ignored.
+          - flush_limit_bytes: Flush immediately after flush_limit_bytes bytes are added.
+          - flow_control_max_mutation_count: Maximum number of inflight mutations.
+          - flow_control_max_bytes: Maximum number of inflight bytes.
+          - batch_operation_timeout: timeout for each mutate_rows operation, in seconds. If None,
+              table default_operation_timeout will be used
+          - batch_per_request_timeout: timeout for each individual request, in seconds. If None,
+              table default_per_request_timeout will be used
+        """
+        self._operation_timeout: float = (
+            batch_operation_timeout or table.default_operation_timeout
+        )
+        self._per_request_timeout: float = (
+            batch_per_request_timeout
+            or table.default_per_request_timeout
+            or self._operation_timeout
+        )
+        if self._operation_timeout <= 0:
+            raise ValueError("batch_operation_timeout must be greater than 0")
+        if self._per_request_timeout <= 0:
+            raise ValueError("batch_per_request_timeout must be greater than 0")
+        if self._per_request_timeout > self._operation_timeout:
+            raise ValueError(
+                "batch_per_request_timeout must be less than batch_operation_timeout"
+            )
+        self.closed: bool = False
+        self._table = table
+        self._staged_entries: list[RowMutationEntry] = []
+        self._staged_count, self._staged_bytes = 0, 0
+        self._flow_control = _FlowControl(
+            flow_control_max_mutation_count, flow_control_max_bytes
+        )
+        self._flush_limit_bytes = flush_limit_bytes
+        self._flush_limit_count = (
+            flush_limit_mutation_count
+            if flush_limit_mutation_count is not None
+            else float("inf")
+        )
+        self._flush_timer = self._start_flush_timer(flush_interval)
+        self._flush_jobs: set[asyncio.Future[None]] = set()
+        # MutationExceptionGroup reports number of successful entries along with failures
+        self._entries_processed_since_last_raise: int = 0
+        self._exceptions_since_last_raise: int = 0
+        # keep track of the first and last _exception_list_limit exceptions
+        self._exception_list_limit: int = 10
+        self._oldest_exceptions: list[Exception] = []
+        self._newest_exceptions: list[Exception] = []
```

Comment:
should this be a deque?

Reply:
I don't think it should make too much of a difference for this use-case, but I changed to deque in case that changes